### ISSUE TITLE
Configuration order and logging

### ISSUE THREAD

**Issue by battlmonstr** [Instabug/Instabug-iOS#224]
### Steps to Reproduce the Problem

Following instructions I have added this code:

```
    [Instabug setSDKDebugLogsLevel:IBGSDKDebugLogsLevelNone];
    [Instabug setCrashReportingEnabled:NO];
    [Instabug startWithToken:@"xxx" invocationEvent:IBGInvocationEventNone];
```

### Expected Behavior

* no logging
* no crash reporting

### Actual Behavior

Every configuration call before "startWithToken" produces a warning message:

> [Instabug] Please make sure to initialize Instabug to be able to use the SDK.

It also affects methods like setPushNotificationsEnabled, setIntroMessageEnabled, and other setXXX configuration calls.

I get something like this in console:

[Instabug] Please make sure to initialize Instabug to be able to use the SDK.[Instabug] Please make sure to initialize Instabug to be able to use the SDK.[Instabug] Please make sure to initialize Instabug to be able to use the SDK.[Instabug] Please make sure to initialize Instabug to be able to use the SDK.[Instabug] Please make sure to initialize Instabug to be able to use the SDK.[Instabug] Please make sure to initialize Instabug to be able to use the SDK.[Instabug] Please make sure to initialize Instabug to be able to use the SDK.

Those messages are unexpected.

If I set IBGSDKDebugLogsLevelDebug, it could be printed, but please fix the newline.

### SDK Version

7.7.1

**Comment by ZiadAyman** [comment]
hey @battlmonstr We are on it!
Will notify you back here once everything is in place.

**Comment by ZiadAyman** [comment]
Hey @battlmonstr 
After investigation, it seems that the problem comes from the order of the APIs initialized. 
Every other API should be initialized after the SDK initialization API `  [Instabug startWithToken:@"xxx" invocationEvent:IBGInvocationEventNone]`
So please give it a try with reversing the order, initializing the SDK first and tell me if you still see the problem.

As for the error line breaks, we'll make sure to fix this in our coming release :)

Please tell me if anything is not clear.

**Comment by danielvidexio** [comment]
Thanks for investigation, but your recommendation is confusing to me. My programming experience hunch says that first you need to configure something, then start it. Your explanation tells me to do it the other way around. If I do so, can you guarantee that the SDK won't initialize and start any unrequested features, won't send out any data to support disabled features? I'm talking about setPushNotificationsEnabled:NO, setIntroMessageEnabled:NO among the others. Will you "restart" it after each "setXXX" call?

Another problem is that setSDKDebugLogsLevel:IBGSDKDebugLogsLevelNone must disable logging. How can I guarantee that startWithToken is not logging anything if I don't call setSDKDebugLogsLevel in advance?

**Comment by battlmonstr** [comment]
Thanks for investigation, but your recommendation is confusing to me. My programming experience hunch says that first you need to configure something, then start it. Your explanation tells me to do it the other way around. If I do so, can you guarantee that the SDK won't initialize and start any unrequested features, won't send out any data to support disabled features? I'm talking about setPushNotificationsEnabled:NO, setIntroMessageEnabled:NO among the others. Will you "restart" it after each "setXXX" call?

Another problem is that setSDKDebugLogsLevel:IBGSDKDebugLogsLevelNone must disable logging. How can I guarantee that startWithToken is not logging anything if I don't call setSDKDebugLogsLevel in advance?

**Comment by ZiadAyman** [comment]
Okay I agree it might be a bit confusing yes. 
Think of `[Instabug startWithToken:@"xxx" invocationEvent:IBGInvocationEventNone];` as an initialization step. Its not the actual start of invoking everything or logging any kind of logs. 
The only logs you can get from the initialization API will look like `Starting version 7.6 (8855) of the Instabug SDK` which logs the SDK version and the number of the build. 
Disabling logs right after this will prevent the SDK from logging anything. 

So, its all about getting the SDK initialized so any other API performs its action later. 
Clearer?

**Comment by battlmonstr** [comment]
> "startWithToken" as an initialization step

Of course it is an initialization step, as well as setSDKDebugLogsLevel (and other setXXX) is also an initialization step. But "startWithToken" is a special in a sense that it allows performing actions.

One of the actions is to print a log message "Starting version 7.6...". By calling setSDKDebugLogsLevel at first, I wish to avoid printing that line (and it works well by the way).

> getting the SDK initialized so any other API performs its action later

An action can be "show the intro message". This action might be triggered after SDK is initialized as you say. The SDK is initialized by "startWithToken", so there's no other way for me to be sure that this action is disabled, unless I call setIntroMessageEnabled in advance.

This is common sense. It is logical to have an initialization order like so:

```
[Instabug setXXX...]
[Instabug setYYY...]
[Instabug setZZZ...]
[Instabug start...]
```

Thank you for checking and ensuring that it will work the other way around.

All I'm saying is that 1) this order of initialization "set" then "start" is more intuitive 2) "Please make sure to initialize Instabug" message should not be shown after setSDKDebugLogsLevel. No matter what the order is.

**Comment by ZiadAyman** [comment]
What you say makes complete sense here. 
We will be working to enable setting `setSDKDebugLogsLevel` before initializing the SDK so you can set it to NO, completely silent the SDK and stop receiving any logs from Instabug. 
Will notify you back once the change is ready!

**Comment by ZiadAyman** [comment]
hey @battlmonstr 
We are done with the change here and now `setSDKDebugLogsLevel` API could be initialized before initializing the SDK so in case you need to completely silent the SDK you can use 
`[Instabug setSDKDebugLogsLevel:IBGSDKDebugLogsLevelNone];` before 
`[Instabug startWithToken:@"xxx" invocationEvent:IBGInvocationEventNone];`
The fix will be released earlier next week, are you okay with this or you need a custom build for the fix today? Let me know please.

**Comment by battlmonstr** [comment]
Thanks a lot for your effort.

I'm ok to wait for the release. I'll be able to test it in about 1-2 weeks time frame. Feel free to close the issue, I'll find it and comment if needed later.

It's very nice to see people who care about their product.

**Comment by ZiadAyman** [comment]
Thank you so much for the sweet words! @battlmonstr 
This will be released in our coming release next Sunday the 21st of Jan. 
I will notify you here one more time next Sunday after the release and then close it :)

**Comment by tadija** [comment]
Hi, is this issue fixed in the latest 7.8.1 release?

I still see this message `"Please make sure to initialize Instabug to be able to use the SDK."` if I use `Instabug.setSDKDebugLogsLevel(.error)` and not if I use `Instabug.setSDKDebugLogsLevel(.none)`, but then I wouldn't be informed in case that some errors really happen.

**Comment by ZiadAyman** [comment]
Hey @tadija 
Actually, we just reproduced this problem yes. `[Instabug startWithToken:@"xxx" invocationEvent:IBGInvocationEventNone];` logs this message. We are fixing it and releasing this tomorrow!

As for initializing `Instabug.setSDKDebugLogsLevel(.none)` you can already initialize it before initializing the SDK. Its now the only API that might get executed before initializing the SDK in case you need to completely silent your SDK. 

Please let me know if anything is not clear from my side.

**Comment by ZiadAyman** [comment]
@battlmonstr @tadija Everything now is fixed and released on iOS SDK version 7.8.2.
Upgrade and let me know if anything is not in place!

**Comment by tadija** [comment]
Hi @ZiadAyman, I can confirm that this is fixed now, thanks!
